src/ui/hexedit.cc

```diff
@@ -1341,24 +1346,61 @@ void HexEdit::saveChunkToFile(const QString& path) {
   saveDataToFile(byte_offset, size, path);
 }
 
+/**
+ * The data is saved to a temporary file and then moved to the destination,
+ * so that the previous contents wouldn't be destroyed if something went wrong
+ * (e.g. connection with server was broken and couldn't load data to save).
+ */
 void HexEdit::saveDataToFile(qint64 byte_offset, qint64 size,
                              const QString& path) {
   if (byte_offset + size > dataBytesCount_) {
     size = dataBytesCount_ - byte_offset;
   }
+  auto data = edit_engine_.bytesValues(byte_offset, size);
 
-  auto data_to_save = edit_engine_.bytesValues(byte_offset, size);
+  QString tmp_path = path + ".tmp." + util::generateRandomUppercaseText(12);
 
-  QFile file(path);
+  QFile file(tmp_path);
   if (!file.open(QIODevice::WriteOnly)) {
-    QMessageBox::information(this, tr("Unable to open file"),
-                             file.errorString());
+    QMessageBox::warning(this, tr("Unable to create temporary file"),
+                         tr("Unable to create temporary file '%1': %2")
+                             .arg(tmp_path)
+                             .arg(file.errorString()));
     return;
   }
 
-  QDataStream stream(&file);
-  stream.writeRawData(reinterpret_cast<const char*>(data_to_save.rawData()),
-                      static_cast<int>(data_to_save.octets()));
+  auto remove_tmp_file = [&file]() {
+    if (!file.remove()) {
+      QMessageBox::warning(
+          this, tr("Unable to remove temporary file"),
+          tr("Unable to remove temporary file '%1'").arg(tmp_path));
+    }
+  };
+
+  if (file.write(reinterpret_cast<const char*>(data.rawData()),
+                 static_cast<int>(data.octets())) == -1) {
+    QMessageBox::warning(
+        this, tr("Error writing to temporary file"),
+        tr("Error writing to temporary file '%1'.").arg(tmp_path));
+    remove_tmp_file();
+    return;
+  }
+  if (QFile::exists(path)) {
+    if (!QFile::remove(path)) {
+      QMessageBox::warning(this, tr("Error"),
+                           tr("File \'%1\' cannot be replaced.").arg(path));
+      remove_tmp_file();
+      return;
+    }
+  }
+  if (!file.rename(path)) {
+    QMessageBox::warning(this, tr("Error"),
+                         tr("Error moving temporary file '%1' to '%2'.")
+                             .arg(tmp_path)
+                             .arg(path));
+    remove_tmp_file();
+    return;
+  }
 }
```

Comment:
`QFile` destructor can fail (when flushing buffers), we have to close it manually and check error code.

Reply:
But `file.remove()` and `file.rename(path)` always close the file.

Reply:
I'm not sure if `QFile::rename` reopens the file after renaming, we should check it.

Reply:
The documentation is unclear, so I checked the source and it seems that `QFile::rename` always leaves the object in closed state.